lib/semantics/check-do-stmt.cc

```diff
@@ -344,41 +372,101 @@ static CS GatherReferencesFromExpression(const parser::Expr &expression) {
   }
 }
 
-// Find a canonical DO CONCURRENT and enforce semantics checks on its body
-class DoConcurrentContext {
+// Find a DO statement and enforce semantics checks on its body
+class DoStmtContext {
 public:
-  DoConcurrentContext(SemanticsContext &context)
-    : messages_{context.messages()} {}
-
-  bool operator==(const DoConcurrentContext &x) const { return this == &x; }
+  DoStmtContext(SemanticsContext &context) : context_{context} {}
 
   void Check(const parser::DoConstruct &doConstruct) {
+    if (doConstruct.IsDoConcurrent()) {
+      CheckDoConcurrent(doConstruct);
+      return;
+    }
+    if (doConstruct.IsDoNormal()) {
+      CheckDoNormal(doConstruct);
+      return;
+    }
+    // TODO: handle the other cases
+  }
+
+private:
+  using Bounds = parser::LoopControl::Bounds;
+
+  const Bounds &GetBounds(const parser::DoConstruct &doConstruct) {
+    auto &loopControl{doConstruct.GetLoopControl().value()};
+    return std::get<Bounds>(loopControl.u);
+  }
+
+  void CheckDoControl(parser::CharBlock sourceLocation, bool isReal) {
+    bool warn = context_.warnOnNonstandardUsage() ||
+        context_.ShouldWarn(parser::LanguageFeature::RealDoControls);
+    bool enabled = context_.IsEnabled(parser::LanguageFeature::RealDoControls);
+    if (isReal && enabled && !(warn)) {
+      // No messages for the default case
+    } else if (isReal && enabled && warn && !context_.warningsAreErrors()) {
+      context_.Say(sourceLocation, "DO controls should be INTEGER"_en_US);
+    } else {
+      context_.Say(sourceLocation, "DO controls should be INTEGER"_err_en_US);
+    }
+  }
+
+  void CheckDoVariable(const parser::ScalarName &scalarName) {
+    const DeclTypeSpec *symType{scalarName.thing.symbol->GetType()};
+    if (symType->IsNumeric(TypeCategory::Integer)) {
+      return;  // No warnings or errors for INTEGER
+    }
+    const parser::CharBlock &sourceLocation{scalarName.thing.source};
+    CheckDoControl(sourceLocation, symType->IsNumeric(TypeCategory::Real));
+  }
+
+  void CheckDoExpression(const parser::ScalarExpr &scalarExpression) {
+    const evaluate::Expr<evaluate::SomeType> *expr{GetExpr(scalarExpression)};
```

Comment:
You should also check that `expr != nullptr`, it may be null if the expression analysis failed. For instance, the following program segfault this PR:

```
program test
  integer i
  ! Expr may be invalid, like here (literal cannot be represented)
  do i = -2147483648_4, 10_4, 3_4
    print *, "segfault"
  end do
end program
```

Reply:
Again, thanks for catching this.  You're a genius!  I'll add a fix and a test.